[DPA-1519]: feat(mcms): update SetConfig changeset to support solana

graham-chainlink opened the pull request:
Update SetConfig changeset to support solana.

JIRA: https://smartcontract-it.atlassian.net/browse/DPA-1519

graham-chainlink commented on deployment/common/changeset/set_config_mcms.go:
```diff
@@ -332,27 +389,82 @@ func SetConfigMCMSV2(e deployment.Environment, cfg MCMSConfigV2) (deployment.Cha
+
+			var transactions []mcmstypes.Transaction
+			if useMCMS {
+				// is this correct?
```
Is this right? Do i convert all the instructions into a single transaction?

gustavogama-cll replied:
looks correct, but you're not converting the instructions into a single transaction. Each instruction is converted into a separate transaction.

gustavogama-cll replied:
I would extract this to part into a separate method. It pollutes this function, imo. Maybe `setConfigPerRoleSolana` should return mcms transactions, not instructions.

graham-chainlink replied:
> I would extract this to part into a separate method. It pollutes this function, imo. Maybe setConfigPerRoleSolana should return mcms transactions, not instructions.

yes thats the plan, wanted to check if i am doing the instructions conversion correctly.

> looks correct, but you're not converting the instructions into a single transaction. Each instruction is converted into a separate transaction.

ah i see, let me dig around, trying to see how to combine multiple instructions in to a single transaction.

graham-chainlink commented on deployment/common/changeset/set_config_mcms.go:
```diff
@@ -332,27 +389,82 @@ func SetConfigMCMSV2(e deployment.Environment, cfg MCMSConfigV2) (deployment.Cha
-			batch := addTxsToProposalBatchV2(setConfigTxsChain, chainSelector, *state)
-			batches = append(batches, batch)
+
+		switch family {
```
Should i separate EVM and Solana into its own function and put them in `internal`?

gustavogama-cll replied:
I'd say so. That's how it was suggested that we handle the deployment changeset.

graham-chainlink replied:
I moved solana out but kept the EVM in the mean time to keep the change small, can move evm out when we come back and delete the v1 stuff

gustavogama-cll replied:
So... let me vent a little to get we all thinking about a solution for this.

The more I work on the common changesets, the less I like the idea of the `internal` subpackage. It does not seem to be providing any benefits, tbh. Quite the contrary, we have to go through lots of detours in order to avoid import cycle issues.

CCIP doesn't use the internal package. Keystone does seem to use it, but I don't know how good the actual implementation is. I'm not sure... maybe there's a better way to rearrange things and keep the internal subpackage -- create some explicit rules about what should go where, and how to compose things. But, right now, I feel that the current package structure is unintuitive and ends up slowing us down.

Maybe there's already a plan or official guideline on how to structure this. If anyone knows of such a thing of has any concrete ideas, please let me know.

cc @graham-chainlink @ecPablo @jkongie

ecPablo replied:
I +1 removing the internal pkg. Originally the old structure had an internal package when everything was EVM https://github.com/smartcontractkit/chainlink/tree/8af9a9097549bfb36a12f1e7e245f0a051dc1dbf/deployment/common/changeset so we just kept it in an effort to keep a similar structure to the existing one per chain family. But now that we have full ownership of this I think we can simplify the pkg structure and simply make private any functions we don't want imported.

graham-chainlink replied:
alright sounds good to me, i will update the PR.

graham-chainlink replied:
Updated! No longer using the internal folder

gustavogama-cll commented on deployment/common/changeset/set_config_mcms_test.go:
I think we should add solana-specific checks to TestSetConfigMCMSV2Variants

graham-chainlink replied:
defintely!

graham-chainlink commented on deployment/common/changeset/set_config_mcms_test.go:
```diff
@@ -167,52 +170,36 @@ func TestSetConfigMCMSV2Variants(t *testing.T) {
 	for _, tc := range []struct {
 		name       string
-		changeSets func(mcmsState *commonchangeset.MCMSWithTimelockState, chainSel uint64, cfgProp, cfgCancel, cfgBypass mcmstypes.Config) []commonchangeset.ConfiguredChangeSet
+		changeSets func(timelockConfig commonchangeset.TransferToMCMSWithTimelockConfig, cfgs map[uint64]commontypes.ConfigPerRoleV2) []commonchangeset.ConfiguredChangeSet
```
What would be a good way to test this for solana? We dont have an equivalent `TransferToMCMSWithTimelockConfig` for solana.

gustavogama-cll commented on deployment/common/changeset/set_config_mcms.go:
```diff
@@ -387,3 +422,74 @@ func addTxsToProposalBatchV2(setConfigTxsChain setConfigTxs, chainSelector uint6
+	if err != nil {
+		return nil, err
+	}
+	bypasserTxs, err := setConfigForRole(e, chain, cfg.Bypasser, bypasserAddress, string(commontypes.CancellerManyChainMultisig), useMCMS)
```
```suggestion
	bypasserTxs, err := setConfigForRole(e, chain, cfg.Bypasser, bypasserAddress, string(commontypes.BypasserManyChainMultisig), useMCMS)
```

graham-chainlink replied:
nice pickup!

graham-chainlink committed:
feat(mcms): update SetConfig changeset to support solana

Update SetConfig changeset to support solana.

JIRA: https://smartcontract-it.atlassian.net/browse/DPA-1519
